python/ray/serve/_private/router.py

```diff
@@ -539,107 +599,134 @@ async def choose_two_replicas_with_backoff(
+                    tried_same_az = True
                 else:
-                    candidate_replica_ids = (
-                        self._get_candidate_replica_ids_for_multiplexed_model_id(
-                            request_metadata.multiplexed_model_id,
-                            get_from_all_replicas=True,
-                        )
+                    # On subsequent iterations or when there are no replicas on the same
+                    # node or AZ, consider all available replicas.
+                    candidate_replica_ids = self._replica_id_set
+
+                if candidate_replica_ids:
+                    chosen_ids = random.sample(
+                        list(candidate_replica_ids),
+                        k=min(2, len(candidate_replica_ids)),
                     )
-            elif (
-                self._prefer_local_node_routing
-                and not tried_same_node
-                and len(self._colocated_replica_ids[LocalityScope.NODE]) > 0
-            ):
-                # Attempt to schedule requests to replicas on the same node at most once
-                candidate_replica_ids = self._colocated_replica_ids[LocalityScope.NODE]
-                tried_same_node = True
-            elif (
-                self._prefer_local_az_routing
-                and not tried_same_az
-                and len(self._colocated_replica_ids[LocalityScope.AVAILABILITY_ZONE])
-                > 0
-            ):
-                # Attempt to schedule requests to replicas in the same AZ at most once
-                candidate_replica_ids = self._colocated_replica_ids[
-                    LocalityScope.AVAILABILITY_ZONE
-                ]
-                tried_same_az = True
-            else:
-                # On subsequent iterations or when there are no replicas on the same
-                # node or AZ, consider all available replicas.
-                candidate_replica_ids = self._replica_id_set
+                    yield [self._replicas[chosen_id] for chosen_id in chosen_ids]
 
-            if candidate_replica_ids:
-                chosen_ids = random.sample(
-                    list(candidate_replica_ids), k=min(2, len(candidate_replica_ids))
-                )
-                yield [self._replicas[chosen_id] for chosen_id in chosen_ids]
+                if not entered_backoff:
+                    entered_backoff = True
+                    self.num_scheduling_tasks_in_backoff += 1
+                    self.num_scheduling_tasks_in_backoff_gauge.set(
+                        self.num_scheduling_tasks_in_backoff
+                    )
 
-            await asyncio.sleep(self.backoff_sequence_s[backoff_index])
-            backoff_index = min(backoff_index + 1, len(self.backoff_sequence_s) - 1)
+                await asyncio.sleep(self.backoff_sequence_s[backoff_index])
+                backoff_index = min(backoff_index + 1, len(self.backoff_sequence_s) - 1)
+        finally:
+            if entered_backoff:
```

Comment:
I don't think we need all that complexity to track retries: instead we can just emit number of retries and that should be sufficient for us to track (we can't track per-actor metrics anyway)

Reply:
What do you mean by `emit number of retries`?

Reply:
Chatted offline:

 - For the current design we actually want to track a number of scheduling tasks
 - We will revisit in a follow-up decoupling scheduling from replicas assignment (and start tracking number of retries instead of number of tasks)